Update 'qualifiers' rule in core spec #382 (#398)
package-url/purl-spec: PURL-SPECIFICATION.rst

> Package URL specification v1.0.X (from qualifiers PR)

Reviewer: is this change of the title intended for meeging?

Author: No; thank you for spotting this @jkowalleck -- it's a temp header I use to check how these .rst pages look on a local test Sphinx RTD before pushing and viewing their GH display.  ;-)  Will remove in the next push along with changes discussed below.

Author: Fixed.

  - The ``type`` MUST start with an ASCII letter.
  - The ``type`` MUST NOT be percent-encoded.
  - The ``type`` is case insensitive. The canonical form is lowercase.

>   - ``purl`` parsers MUST return an error when the ``type`` contains a prohibited

Reviewer: this change is not in the scope of the https://github.com/package-url/purl-spec/issues/382

Anyway, :+1: this is in accordance with the existing spec.

https://github.com/package-url/purl-spec/blob/8040ff0be50f0c5b1986b1a0947bd539f5405fc4/PURL-SPECIFICATION.rst?plain=1#L245

Author: Thanks @jkowalleck -- I forgot my own advice to keep PRs narrowly focused.  Will remove in the next iteration and address in a separate PR.  :+1:

Author: Backed out the type update.

- **qualifiers**:

  - The ``qualifiers`` component MUST be prefixed by a '?' separator when not empty.
  - The '?' separator is not part of the ``qualifiers`` component.
  - The ``qualifiers`` component is a query string composed of one or more ``key=value``
    pairs.   Multiple ``key=value`` pairs MUST be separated by an ampersand '&'.
    A ``key`` and ``value`` MUST be separated by the equal '=' character.
  - Neither the '&' nor the '=' separator is part of the ``key`` or the ``value``.
  - Each ``key`` MUST be unique among the keys of the ``qualifiers`` string.
  - A ``value`` MUST NOT be an empty string: a ``key=value`` pair with an empty ``value``

>     is the same as no ``key=value`` pair at all for this ``key``.

Reviewer 2: ```suggestion
    is the same as no ``key=value`` pair existed for this ``key``.
```

Author: Good clarification -- done.

- **qualifiers**:

  - The ``qualifiers`` component MUST be prefixed by an unencoded question
    mark '?' separator when not empty.  This '?' separator is not part of the
    ``qualifiers`` component.

>   - The ``qualifiers`` component is a query string composed of one or more

Reviewer: ```suggestion
  - The ``qualifiers`` component is a sequence of one or more
```

Reviewer 2: This make sense, as the query string references does not bring anything special (though this has specific meaning in the URL specs).
What about going even simpler, as a sequence is also a new term:
```suggestion
  - The ``qualifiers`` component is composed of one or more
```

Reviewer: yea, the "sequence" would mean that things are in a certain order - and this is irrelevant.
lets just use the " ... is composed of one or more ..." prase

    ``key=value`` pairs.  Multiple ``key=value`` pairs MUST be separated by an
    unencoded ampersand '&'.  This '&' separator is not part of the

>     ``qualifiers`` component.

Reviewer: ```suggestion
    ``qualifiers``' sub-components.
```

Reviewer 2: @jkowalleck Let's not introduce a sub-component concept.

Author: Done.

  - A ``key`` and ``value`` MUST be separated by the unencoded equal sign '='
    character.  This '=' separator is not part of the ``key`` or ``value``.
  - A ``value`` MUST NOT be an empty string: a ``key=value`` pair with an
    empty ``value`` is the same as if no ``key=value`` pair exists for this
    ``key``.

  - For each ``key=value`` pair:

    - The ``key`` MUST be composed only of lowercase ASCII letters and numbers,
      period '.', dash '-' and underscore '_'.
    - A ``key`` MUST start with an ASCII letter.
    - A ``key`` MUST NOT be percent-encoded.
    - Each ``key`` MUST be unique among all the keys of the ``qualifiers``

>       component.

Reviewer: ```suggestion
      sub-component.
```

Reviewer: see https://github.com/package-url/purl-spec/pull/398#discussion_r2019131319

Reviewer: with current changes applied, i see no longer a need to change this. 
closd.

- **qualifiers**:

  - The ``qualifiers`` component MUST be prefixed by a '?' separator when not empty.
  - The '?' separator is not part of the ``qualifiers`` component.
  - The ``qualifiers`` component is a query string composed of one or more ``key=value``
    pairs.   Multiple ``key=value`` pairs MUST be separated by an ampersand '&'.
    A ``key`` and ``value`` MUST be separated by the equal '=' character.
  - Neither the '&' nor the '=' separator is part of the ``key`` or the ``value``.
  - Each ``key`` MUST be unique among the keys of the ``qualifiers`` string.
  - A ``value`` MUST NOT be an empty string: a ``key=value`` pair with an empty ``value``
    is the same as no ``key=value`` pair at all for this ``key``.

  - For each ``key=value`` pair:

    - The ``key`` MUST be composed only of ASCII letters and numbers, '.', '-' and
      '_' (period, dash and underscore).
    - A ``key`` MUST start with an ASCII letter.
    - A ``key`` MUST NOT be percent-encoded.
    - A ``key`` is case insensitive. The canonical form is lowercase.

>     - A ``value`` MAY be composed of any ASCII or non-ASCII character, and

Author: @pombredanne The "Character encoding" section already refers to ASCII and non-ASCII -- that's where my references come from.  https://github.com/package-url/purl-spec/blob/main/PURL-SPECIFICATION.rst#character-encoding

>     - A ``value`` MAY be composed of any ASCII or non-ASCII character, and

Author: @pombredanne If I'm reading the "blame" data correctly, this language was already in the Character encoding section when you added your 2017-11-13 commit:

```
For clarity and simplicity a `purl` is always an ASCII string.
To ensure that there is no ambiguity when parsing a `purl`, separator
characters and non-ASCII characters must be UTF-encoded and then
percent-encoded as defined at::
```

See https://github.com/package-url/purl-spec/commit/840410b684ee31ab7cdd0c66cb5fe19f11d6cc7e

>     - A ``value`` MAY be composed of any ASCII or non-ASCII character, and

Author: @pombredanne  FWIW:

- RFC 3986 has 15 references to "ASCII", including 9 to "US-ASCII" (referring to DEC 0-127) and 4 to "non-ASCII" (e.g., "non-ASCII data", "non-ASCII registered names", "[n]on-ASCII characters").  
- The "Normative Reference" for US-ASCII (RFC 3986 section 10.1) is "American National Standards Institute, "Coded Character Set -- 7-bit American Standard Code for Information Interchange", ANSI X3.4, 1986."  
- There is no definition of "non-ASCII".

(https://datatracker.ietf.org/doc/html/rfc3986)

>     - A ``value`` MAY be composed of any ASCII or non-ASCII character, and

Reviewer 2: @johnmhoran Thanks for all the details. Here is my take:
```suggestion
    - A ``value`` MAY be composed of any character. A value MUST be percent-encoded as described in the "Character encoding" section.
```
Then we can clarify the "Character encoding" in that section, not here.

>     - A ``value`` MAY be composed of any ASCII or non-ASCII character, and

Author: @pombredanne Done.  As a result of this change, with the current language in "Character encoding", all characters used in a `qualifiers` `value` must be percent-encoded except:

  (1) an unreserved character as defined in RFC 3986 section 2.3 (https://datatracker.ietf.org/doc/html/rfc3986#section-2.3), or

  (2) expressly defined in this PURL-SPECIFICATION.rst as a ``purl`` separator (and only when used as such a separator).

That would mean that with the current "Character encoding" language, the colon ':', for example, MUST be percent-encoded in the `qualifiers` `value`.  So, are you contemplating:

- adding exceptions to the `qualifiers` section (that's the language I added and at your request just removed), or 
- adding exceptions to the "Character encoding" section (presumably this would include the exceptions all components that have 1+ exceptions), or 
- rewriting the default `MUST be percent-encoded` in the "Character encoding" section, or
- some other approach?

We could also just direct users to read the wikipedia percent-encoding page and/or RFC 3986, but both contain ambiguities and we'd need to keep fielding new issues and PRs from puzzled users/potential users of the purl spec.  I vote for clarity instead.

>     - A ``value`` MAY be composed of any ASCII or non-ASCII character, and

Reviewer 3: Encoding colon in any part of the PURL (or any other currently unencoded character) changes the canonical form in a way that will cause problems for some users and shouldn't be done without good reason. One of the reasons I originally got involved with PURL implementations was related to problems caused by inconsistent colon encoding across the "canonical" forms of different implementations.

>     - A ``value`` MAY be composed of any ASCII or non-ASCII character, and

Author: Thanks @matt-phylum.  I've just pushed my updates in response to several of @pombredanne's and other comments.  

Can you please propose language for the `qualifiers` section and the "Character encoding" section that will (1) accomplish your goals and (2) add sufficient detail and breadth to enable a user to figure out the encoding of any permitted character without needing to open an issue?

      MUST be percent-encoded as described in the "Character encoding"
      section below[.] [, except that:

>       - *list exceptions to the updated "Character encoding" rules, e.g., if a

Reviewer: TODO // to be decided
> colon ':' does not need to, or SHOULD NOT, or MUST NOT, be percent-encoded.*]

Reviewer: If i remember correclty,
frome existing rules, there is no "MUST NOT".
there was a rule that wllowes to percent-encode everything, expcept the well-known delimiters in their expllicitlocations. 

I personally liked that, since this made it easy tp program/codify the rules - when in dounbt, just percent-encode.
and in general: percent-decode everything.

Reviewer 3: Yes. It's easy to implement non-canonical formatting if you're allowed to percent-encode additional characters, and it's easier to implement a parser if you are just decoding and not checking if there were extra characters that shouldn't have been encoded, but the implementation will fail the test suite and users that do string comparisons on PURLs will get false negatives when comparing the output of the implementation with the output of another implementation.

Should the canonical format stuff be separated out somehow? You MUST NOT encode : ever in the canonical form, but if you don't care about the canonical form it's a SHOULD NOT encode except for the first colon which remains a MUST NOT because the parsers really don't care about unnecessary encoding. It'd close a lot of implementation bugs where characters are not being encoded exactly right, but I worry about confusion and usability problems if people are using PURLs produced from multiple implementations. AFAIK URL does not have the same canonical form feature and instead you're expected to parse URL before comparing it. It'd be a much bigger change than this PR.

Author: imho that language is not clear because line 231 refers expressly to the colon separator, not the literal colon character wherever it might be used in the purl.  This could easily be clarified -- IF that's the intent.  Is it?  All we need is some agreement.  🙂

- **qualifiers**:

  - The ``qualifiers`` component MUST be prefixed by a '?' separator when not empty.
  - The '?' separator is not part of the ``qualifiers`` component.
  - The ``qualifiers`` component is a query string composed of one or more ``key=value``

>     pairs, each of which is separated by an ampersand '&'. A ``key`` and ``value``

Author: Change made.

>     are separated by the equal '=' character.

Author: Change made.

>   - The '&' separator is not part of the ``key`` or the ``value``.

Author: Done.

>   - The '=' separator is not part of the ``key`` or the ``value``.

Author: Done.

>   - The ``key`` MUST be unique among the keys of the ``qualifiers`` string.

Author: Change made.

>   - The ``value`` MUST NOT be an empty string: a ``key=value`` pair with an empty ``value``

Author: Done.

    is the same as no ``key=value`` pair at all for this ``key``.

  - For each ``key=value`` pair:

    - ``key``

      - The ``key`` MUST be composed only of ASCII letters and numbers, '.', '-' and
        '_' (period, dash and underscore).

>       - A ``key`` MUST start with an ASCII letter.

Reviewer: > A ``key`` MUST start with an ASCII letter.

this is a change.
old: 
> A ``key`` cannot start with a number

the new spec would prevent keys like `_foo`, `-bar`, `.bazz`

Author: Given the comments thus far ^, I'm keeping as is unless/until I hear otherwise.

>       - A ``key`` MUST NOT be percent-encoded.

Reviewer: > A ``key`` MUST NOT be percent-encoded.

I think this is wrong, and against other existing spec
https://github.com/package-url/purl-spec/blob/8040ff0be50f0c5b1986b1a0947bd539f5405fc4/PURL-SPECIFICATION.rst?plain=1#L245-L248

I think a key can be percent-encoded. and at some points, it must be percent encoded.
Otherwise, if percent-encoding MUST NOT happen, then i could not choose certain keys.
examples: `foo&bar[]=bazz` -> `foo%26bar%5B%5D` -- at least the `&` MUST be percent-encoded.

I might be wrong, please help me understand.

Reviewer 3: `foo&bar[]=bazz` is not a valid key according to the preceding rules. The allowed characters shouldn't _require_ percent encoding but I don't see why the spec would forbid percent encoding.

Author: Given the comments thus far ^, I'm keeping as is unless/until I hear otherwise.

Reviewer: +1 - keep as you have it right now

      - A ``key`` is case insensitive. The canonical form is lowercase.

    - ``value``

      - The ``value`` MUST be composed only of the following characters, encoded

>         as described below and in keeping with RFC 3986 Section 2.2:

Reviewer 3: This is confusing and self contradictory. "The `value` MUST be composed only of the following characters", but then the following text defines a set containing all characters.

Author: Thanks @matt-phylum -- I don't think it's confusing but in any event the revised update will be simplified and clarified.

          "If data for a URI component would conflict with a reserved character's
          purpose as a delimiter, then the conflicting data must be percent-encoded
          before the URI is formed."
          https://datatracker.ietf.org/doc/html/rfc3986#section-2.2

        1. **All US-ASCII characters defined as "unreserved"** in RFC 3986 Section 2.3
           (https://datatracker.ietf.org/doc/html/rfc3986#section-2.3):

           .. code-block:: none

               'A'-'Z', 'a'-'z', '0'-'9', '-', '.', '_', '~'

           - These 66 characters do not need to be percent-encoded.

        2. **All US-ASCII characters defined as "sub-delims"**, a subset of
           the "reserved" characters, in RFC 3986 Section 2.2
           (https://datatracker.ietf.org/doc/html/rfc3986#section-2.2):

           .. code-block:: none

>                '!', '$', '&', ''', '(', ')', '*', '+', ',', ';', '='

Reviewer 3: The previous item lists characters that should not be encoded (MUST NOT for canonical PURLs), but this item lists some characters that MUST be encoded and some characters that should not be encoded together.

Author: Thanks, will look forward to your comments on the revised update once I've pushed it.

             as a separator because it splits each key-value pair just once,
             from left-to-right, on the first '=' it encounters, and thus there
             is no conflict:

             .. code-block:: none

                 - For each pair, split the key=value once from left on '=':
                   - The key is the lowercase left side
                   - The value is the percent-decoded right side

        3. **Four additional US-ASCII characters** identified in the "query"
           definition in RFC 3986 Section 3.4 (https://datatracker.ietf.org/doc/html/rfc3986#section-3.4)
           and the "pchar" definition in RFC 3986 Appendix A (https://datatracker.ietf.org/doc/html/rfc3986#appendix-A):

           .. code-block:: none

               ':', '@', '/', '?'

           - The '?' MUST be percent-encoded to avoid being incorrectly parsed
             as a ``qualifiers`` separator -- in the right-to-left parsing
             (see "How to parse a purl string in its components"), an unencoded
             '?' in the ``value`` would be the first '?' encountered by the
             parser and incorrectly treated as a ``qualifiers`` separator.

>            - The other three characters do not need to be percent-encoded.

Reviewer 3: This contradicts preexisting rules: https://github.com/package-url/purl-spec/blob/8040ff0be50f0c5b1986b1a0947bd539f5405fc4/PURL-SPECIFICATION.rst?plain=1#L238-L241

Author: Will be clarified in the update.

        4. **All other US-ASCII characters**.

           .. code-block:: none

               - 33 control characters (ASCII codes 0-31 and 127)

               - the 14 US-ASCII characters not covered in the preceding groups of US-ASCII characters:

                 ' ' [space], '"', '#', '%', '<', '>', '[', '\', ']', '^', '`', '{', '|', '}'

>            - Each of these 47 US-ASCII characters MUST be percent-encoded.

Reviewer 3: This is a change to the PURL canonicalization rules and will cause problems for anyone who is using canonicalized PURLs as keys.

Author: @matt-phylum As you'll see elsewhere in these comments, I've deleted what was intended as a simple placeholder rule hoping people would provide language they wanted.  Failing that, it's been deleted.

        5. **Any character that is not a US-ASCII character**
           (i.e., characters with ASCII code > 127 and non-ASCII characters).

>            - All such characters MUST be UTF-8 encoded and then percent-encoded.

Reviewer 3: This is way too complicated to be one bullet point in the qualifiers section.

The encoding rules are:

- The ASCII control characters 0 through 31 and 127 and any non-ASCII character greater than or equal to 128 MUST always be encoded in any component of a PURL.
- The additional characters ` ` `"` `<` `>` SHOULD always be encoded in any component of a PURL.
- The additional characters `#` `@` `?` `%` MUST be encoded for qualifier values.

Plus should also be encoded to avoid interoperability problems: #261

Author: Will be clarified in the update.

  - ``purl`` parsers MUST return an error when the ``key`` or ``value`` contains

>     a prohibited character.

Reviewer 3: This is incorrect and incompatible with the preexisting spec.

In the unescaped form, no characters are prohibited, so you could have a valid PURL `pkg:generic/name?key=%00` and the parser must handle this (or maybe it returns an error because it's string type can't represent null characters).

In the escaped form, some characters are escaped to avoid problems unrelated to the URL. `http://example.com/?key=a "value"` and `http://example.com/?key=a%20%22value%22` are the same, but if you write `Go to http://example.com/?key=a "value"` or `<a href="http://example.com/?key=a "value"">` it doesn't work. At least I'm pretty sure why it's done. These characters have no meaning to the URL or PURL parsers so whether they are escaped or not doesn't make a difference to the parsing result.

Author: Seems to me that inclusion of a prohibited character could only be "normalized" by removing/discarding such a character, which does not sound to me like mere normalization.  I look to feedback from @pombredanne and others on this point.

Reviewer 3: Maybe I'm misunderstanding. There are no characters that are prohibited in the unencoded form of a qualifier value. In the encoded form, there are characters that are listed above as requiring escaping that don't actually require escaping in order for the PURL to be parsed successfully. If it can be unambiguously parsed and the result can be formatted into a PURL with the same meaning, I don't think that should be a "MUST return an error" condition. The characters do not need to be removed or discarded.

Author: Hmmm.  I must admit that I don't yet have a good understanding of when a violation of the spec should be normalized (allowing the parsing process to continue) vs. treating a spec violation as invalid (i.e., `"is_invalid": true`, raising an error/exception and thus halting the parsing process).  

I'd be interested in reading any authorities/resources you could point me to on this important topic -- and in all events I'll happily defer to whatever approach you, @pombredanne, @jkowalleck, @mprpic and others in the community ultimately agree to.

    in the canonical form
  - Each ``subpath`` segment MUST be a percent-encoded string
  - When percent-decoded, a segment:

    - MUST NOT contain a '/'
    - MUST NOT be any of '..' or '.'
    - MUST NOT be empty

  - The ``subpath`` MUST be interpreted as relative to the root of the package

Character encoding
~~~~~~~~~~~~~~~~~~

> For clarity and simplicity, a ``purl`` is always an ASCII string. To ensure that

Reviewer 2: I would drop "For clarity and simplicity" which is a color commentary that I should not have included in the first place.

Then what about this:
```
A canonical ``purl`` is always an ASCII string composed only of these characters:
``A to Z``, ``a to z``, ``0 to 9`` and these punctuation ``:/@?#%.-_~``
```

... and then we can go on to specify the arcane and bizarre rules of percent encoding for purls for other characters, and that each component defines when and how to use the encoding defined here.

(And we likely should add a reference section further down with something like in RFC3986
   [ASCII]    American National Standards Institute, "Coded Character
              Set -- 7-bit American Standard Code for Information
              Interchange", ANSI X3.4, 1986.

)

Author: @pombredanne I've made these suggested updates as well, except for the detailed work covered by your phrase `and then we can go on to specify the arcane and bizarre rules of percent encoding for purls for other characters`.  We need (and don't yet have) agreement on these details before I can reduce them to text.

I see that you made no changes to the PR's current default of `MUST be percent-encoded unless . . ..`  I'm leaving that as is in the next iteration (which I'll push soon), but I have a suggested change to that default language that I'll add in a new comment once I've pushed that update.

there is no ambiguity when parsing a ``purl``, separator characters and non-ASCII
characters MUST be UTF-encoded and then percent-encoded as defined at
https://en.wikipedia.org/wiki/Percent-encoding and as further defined below.

> Use these rules for percent-encoding and decoding ``purl`` components.  Except

Reviewer 2: Encoding is not specific to components. We are just defining how to encode and decode characters in a string. Sometimes we will refer to encoding for a whole component, some other times, we are stating the segments of a subpath are encoded individually.

Author: Good point -- changed to `percent-encoding and decoding the characters that comprise
a ``purl`` string` -- does that capture your thought @pombredanne ?

as otherwise provided in the "Rules for each ``purl`` component" section above:

- A character used in a ``purl`` component MUST be percent-encoded unless it is:

  (1) an unreserved character as defined in RFC 3986 section 2.3 (https://datatracker.ietf.org/doc/html/rfc3986#section-2.3),

  (2) expressly defined in this PURL-SPECIFICATION.rst as a ``purl`` separator (and only when used as such a separator), or

  (3) expressly permitted in that ``purl`` component.

- All non-ASCII characters MUST be encoded as UTF-8 and then percent-encoded.

- The characters used as ``purl`` separators are listed below.  These characters:

  - MUST NOT be percent-encoded when used as separators.
  - MUST be percent-encoded when not used as separators unless expressly permitted
    by a ``purl`` component.

> .. list-table::

Reviewer 2: I would avoid tables, let's use only bullets

Author: A sensible idea -- done.  ;-)

    in the canonical form
  - Each ``subpath`` segment MUST be a percent-encoded string
  - When percent-decoded, a segment:

    - MUST NOT contain a '/'
    - MUST NOT be any of '..' or '.'
    - MUST NOT be empty

  - The ``subpath`` MUST be interpreted as relative to the root of the package

Character encoding
~~~~~~~~~~~~~~~~~~

A canonical ``purl`` is always an ASCII string composed only of these characters:

- ``A to Z``,
- ``a to z``,
- ``0 to 9`` and

> - the punctuation marks ``:/@?#%.-_~`` .

Reviewer: what about `+` (plus)

we are not www-URL, so a plus is a plus. while a space, which is not in the list of allowed characters, needs to be percent-encoded .... which is declared already. but maybe add also a explicit note how spaces should be handled

Reviewer 3: #261 recommends that a PURL qualifier value should not contain `+` because some implementations do treat it as ` `. But, in the interest of simplifying the encoding, maybe the specified encoding should be the same for qualifier values as everywhere else (whether that's encoded or not).

Reviewer: these implementations are already not respecting the current purl spec.

> - the punctuation marks ``:/@?#%.-_~`` .

Author: Ampersand '&' added.  Good eye.

  underscore '_' and tilde '~').

All other characters MUST be encoded as UTF-8 and then percent-encoded.
In addition, each component specifies its permitted characters and
its percent-encoding rules.

``purl`` separators
-------------------

These ``purl`` separator characters MUST NOT be percent-encoded when used as
``purl`` separators:

- ':' (colon) is the separator between ``scheme`` and ``type``
- '/' (slash) is the separator between ``type``, ``namespace`` and ``name``
- '/' (slash) is the separator between ``subpath`` segments
- '@' (at sign) is the separator between ``name`` and  ``version``
- '?' (question mark) is the separator before ``qualifiers``
- '=' (equals) is the separator between a ``key`` and a ``value`` of a
  ``qualifier``
- '&' (ampersand) is the separator between ``qualifiers`` (each being a
  ``key=value`` pair)
- '#' (number sign) is the separator before ``subpath``

Percent-encoding rules
----------------------

When applying percent-encoding or decoding to a string, use the rules of RFC
3986 section 2 (https://datatracker.ietf.org/doc/html/rfc3986#section-2).

Each component defines when and how to apply percent-encoding and decoding to
its content.

> When percent-encoding is required, all characters MUST be encoded except for

Reviewer 2: This is looking all good to me. IMHO, the last point of discussion left is this sentence. e.g., is colon all we need as a generic rule (leaving aside the per-component rules)?

Reviewer 2: @johnmhoran Let's move this section to a new PR for clarity as discussed in today's Ecma call.

Author: @pombredanne   This has been removed and replaced with the version of "Character encoding" currently in `main`.  I opened a new issue for the current "Character encoding" work: 

- https://github.com/package-url/purl-spec/issues/438

as otherwise provided in the "Rules for each ``purl`` component" section above:

- A character used in a ``purl`` component MUST be percent-encoded unless it is:

  (1) an unreserved character as defined in RFC 3986 section 2.3 (https://datatracker.ietf.org/doc/html/rfc3986#section-2.3),

  (2) expressly defined in this PURL-SPECIFICATION.rst as a ``purl`` separator (and only when used as such a separator), or

  (3) expressly permitted in that ``purl`` component.

- All non-ASCII characters MUST be encoded as UTF-8 and then percent-encoded.

- The characters used as ``purl`` separators are listed below.  These characters:

  - MUST NOT be percent-encoded when used as separators.
  - MUST be percent-encoded when not used as separators unless expressly permitted
    by a ``purl`` component.

.. list-table::
   :widths: 1 4
   :header-rows: 1

   * - Character
     - Use as separator
   * - ':'
     - between ``scheme`` and ``type``
   * - '@'
     - ``version`` prefix
   * - '?'
     - ``qualifiers`` prefix
   * - '#'
     - ``subpath`` prefix
   * - '/'
     - ``type``/``namespace``/``name`` and ``subpath`` segments separator
   * - '='

>      - ``qualifiers`` ``key``/``value`` separator

Reviewer: add  
```rst
   * - '&'
     - ``qualifiers`` ``key=value`` separator
```

Author: Thanks for catching this @jkowalleck -- :+1:  I've added it to the update of this PR that I'll be pushing shortly.
